### ISSUE TITLE
Signature is not valid: Signature was created at XXX and is therefore not yet valid at XXX

### ISSUE THREAD

**Issue by vladojr** [vanitasvitae/pgpainless#444]
Hello, im trying to verify messages that were signed on a computer with time shifted forwards. As this is a realtime system the messages are verified within seconds since creation and the messages fail to verify with error below. It is not possible for me to guarantee that all clients wll have times synced in seconds and thus need way to accept these messages.

Problem can be easily reproduced by shifting the time forwards, createing the message and then verifying it once the time was set back to normal.

I get an error like
Signature is not valid: Signature was created at Wed Jul 24 19:25:38 CEST 2024 and is therefore not yet valid at Wed Jul 24 11:26:47 CEST 2024

I get no such error when using gnupg, neither with bouncypgp I used until now. The behaviour is not necessarily wrong, but it should be possible to override it, either by setting a threshold or disabling this kind of verification.

**Version**

- `pgpainless-core`:

Thank you for great library!

![image](https://github.com/user-attachments/assets/cea10dc9-4796-499f-b03a-bc83365ecf1b)

**Comment by vanitasvitae** [comment]
Hey!
Thank you for your report!

The problem you are describing is quite an edge case. I think it will be hard to come up with a clean change to the API to allow for the creation time check to be disabled.
I guess the best way would be to add a method to `ConsumerOptions` to set a reference date for signature validation.
That way you could set the reference time to the signatures creation time, or take a known delay period into account.

Note though, that GnuPG is pretty much the only other OpenPGP implementation that ignores a future creation time (see https://tests.sequoia-pgp.org/#Detached_signature_with_Subpackets ).

I will put this on the roadmap for a future release, but don't hold your breath :D

**Comment by vladojr** [comment]
Yes, it seems reasonable to configure it using the ConsumerOprions.
Thank you!

**Comment by vanitasvitae** [comment]
This issue just came up in another discussion: https://github.com/pgpainless/pgpainless/discussions/487

You can now work around this problem by adjusting the key creation time during generation, passing in a key creation time that is sufficiently far in the past, as to compensate for localtime differences on separate machines.

**Comment by vladojr** [comment]
I dont see the point here, how does key generation time relate to message signing time? Im already using keys generated in far past.

**Comment by vanitasvitae** [comment]
You are right.
What you want to do is to change the signature creation time during generation, to factor in an offset, taking system dime differences into consideration, e.g. `signatureCreationTime = now - 1h`

I noticed, that this behavior was broken, but will be fixed in 2.0.2.
I added an example test demonstrating the way to deal with system time differences here: https://github.com/pgpainless/pgpainless/commit/3fc5669e56fec0310d9aa5c30be7dac37367650d

**Comment by vladojr** [comment]
The think is that I need to accept messages generated by other systems and libraries and I cannot expect them to adjust the signature creation time backwards.

**Comment by vanitasvitae** [comment]
That's also already possible by adjusting the validity period, especially `verifyNotAfter`. This limit is usually set to `now` in order to invalidate all future signatures, but you can set this to now + 1h for example, to allow verifying signatures made up to 1 hour in the future.

I added a test demonstrating this here: https://github.com/pgpainless/pgpainless/commit/3f7b4920f49d212f8ac82bb524858e75692563d2

**Comment by vladojr** [comment]
Thank you - thats the real fix, I will be trying that out soon!
